Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm

```diff
@@ -0,0 +1,204 @@
+    RetainPtr<WKARPresentationSessionDescriptor> _sessionDescriptor;
+
+    // View state
+    RetainPtr<UIView> _view;
+    WeakObjCPtr<CALayer> _cameraLayer;
+
+    // Camera image
+    RetainPtr<CVPixelBufferRef> _capturedImage;
+}
+
+- (nonnull instancetype)initWithSession:(nonnull ARSession*)session descriptor:(nonnull WKARPresentationSessionDescriptor*)descriptor
+{
+    self = [super init];
+    if (self) {
+        _session = session;
+        _sessionDescriptor = adoptNS([descriptor copy]);
+
+        [self _enterFullscreen];
+    }
+
+    return self;
+}
+
+#pragma mark - WKARPresentationSession
+-(nonnull ARSession *)session {
+    return (ARSession *) _session;
+}
+
+- (NSUInteger)startFrame
+{
+    NOISY_RELEASE_LOG(XR, "%s", __FUNCTION__);
+
+    ARFrame *currentFrame = [_session currentFrame];
+    if (!currentFrame) {
+        RELEASE_LOG(XR, "%s: no frame available", __FUNCTION__);
+        return 0;
+    }
+
+    _capturedImage = currentFrame.capturedImage;
+
+    return 1;
+}
+
+- (void)present
+{
+    NOISY_RELEASE_LOG(XR, "%s", __FUNCTION__);
+
+    [CATransaction begin];
+    {
+        [_cameraLayer setContents:(id) _capturedImage.get()];
+    }
+    [CATransaction commit];
+    [CATransaction flush];
+
+    _capturedImage = nil;
+}
+
+- (void)terminate
+{
+    RELEASE_LOG(XR, "%s", __FUNCTION__);
+}
+
+#pragma mark - UIViewController
+- (void)loadView
+{
+    RELEASE_LOG(XR, "%s", __FUNCTION__);
+    UIView *view = [[UIView alloc] initWithFrame:UIScreen.mainScreen.bounds];
+    self.view = view;
+}
+
+- (void)viewDidLoad
+{
+    RELEASE_LOG(XR, "%s", __FUNCTION__);
+    [super viewDidLoad];
+
+    _cameraLayer = [self.view layer];
+    [_cameraLayer setBackgroundColor:UIColor.whiteColor.CGColor];
+    [_cameraLayer setOpaque:YES];
+}
+
+#if !PLATFORM(VISION)
+- (BOOL)preferStatusBarHidden
+{
+    return YES;
+}
+#endif
+
+- (void)viewWillAppear:(BOOL)animated
+{
+    RELEASE_LOG(XR, "%s", __FUNCTION__);
+    [super viewWillAppear:animated];
+
+    RetainPtr<ARWorldTrackingConfiguration> configuration = adoptNS([WebKit::allocARWorldTrackingConfigurationInstance() init]);
+    [_session runWithConfiguration:configuration.get()];
+}
+
+- (void)viewWillDisappear:(BOOL)animated
+{
+    RELEASE_LOG(XR, "%s", __FUNCTION__);
+    [_session pause];
+    [super viewWillDisappear:animated];
+}
+
+#pragma mark - Private
+- (void)_enterFullscreen
+{
+    ASSERT(RunLoop::isMain());
+    RELEASE_LOG(XR, "%s", __FUNCTION__);
+
+    UIViewController* presentingViewController = [_sessionDescriptor presentingViewController];
+    [presentingViewController presentViewController:self animated:NO completion:^(void) {
+        RELEASE_LOG(XR, "%s: presentViewController complete", __FUNCTION__);
+    }];
+}
+
+@end
+
+id<WKARPresentationSession> createPresesentationSession(ARSession *session, WKARPresentationSessionDescriptor *descriptor)
+{
+    return [[_WKARPresentationSession alloc] initWithSession:session descriptor:descriptor];
+}
```

Comment:
Switching to soft-linking of ARKit means that I couldn't add an extension category to ARSession. I've replaced the method I added to ARSession with a freestanding function.